File: leetcode/1048. Longest String Chain/main.py
class Solution:
    def longestStrChain(self, words: list[str]) -> int:
        words.sort(key=len)

        dp: dict[str, int] = {}
        max_chain = 0

        for word in words:
            dp[word] = 1
            for i in range(len(word)):
                prev = word[:i] + word[i+1:]
                if prev in dp:
                    dp[word] = max(dp[word], dp[prev] + 1)
            max_chain = max(max_chain, dp[word])

        return max_chain
        

File: leetcode/1048. Longest String Chain/main.ts
const longestStrChain = (words: string[]): number => {
  words.sort((a, b) => a.length - b.length);

  const dp: { [key: string]: number } = {};

  let maxChain = 0;

  for (const word of words) {
    dp[word] = 1;
    for (let i = 0; i < word.length; i++) {
      const prev = word.slice(0, i) + word.slice(i + 1);
      if (prev in dp) {
        dp[word] = Math.max(dp[word], dp[prev] + 1);
      }
    }
    maxChain = Math.max(maxChain, dp[word]);
  }

  return maxChain;
};
